examples/esp_idf/hello/main/app_main.c

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2022 Golioth, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "esp_log.h"
+#include "nvs.h"
+#include "shell.h"
+#include "wifi.h"
+#if CONFIG_GOLIOTH_BLE_SERVICE_ENABLED
+#include "ble.h"
+#endif
+#include <golioth/golioth_sys.h>
+#include <golioth/client.h>
+#include <golioth/lightdb_state.h>
+#include <string.h>
+
+#define TAG "hello_espidf"
+
+// Given if/when the we have a connection to Golioth
+static golioth_sys_sem_t _connected_sem;
+
+static void on_client_event(struct golioth_client *client,
+                            enum golioth_client_event event,
+                            void *arg)
+{
+    bool is_connected = (event == GOLIOTH_CLIENT_EVENT_CONNECTED);
+    if (is_connected)
+    {
+        golioth_sys_sem_give(_connected_sem);
+    }
+    GLTH_LOGI(TAG, "Golioth client %s", is_connected ? "connected" : "disconnected");
+}
+
+void app_main(void)
+{
+    int counter = 0;
+
+    // Initialize NVS first. For this example, it is assumed that WiFi and Golioth
+    // PSK credentials are stored in NVS.
+    nvs_init();
+
+#if CONFIG_GOLIOTH_BLE_SERVICE_ENABLED
+    // Initialize BLE stack, which can be used for provisioning WiFi and Golioth credentials
+    ble_init("golioth_esp32");
+#endif
+
+    // Create a background shell/CLI task (type "help" to see a list of supported commands)
+    shell_start();
+
+    // If the credentials haven't been set in NVS, we will wait here for the user
+    // to input them via the shell.
+    if (!nvs_credentials_are_set())
+    {
+        while (1)
+        {
+            golioth_sys_msleep(1000);
+            ESP_LOGW(TAG, "WiFi and golioth credentials are not set");
+            ESP_LOGW(TAG, "Use the shell settings commands to set them, then restart");
+            golioth_sys_msleep(UINT32_MAX);
+        }
+    }
+
+    // Initialize WiFi and wait for it to connect
+    wifi_init(nvs_read_wifi_ssid(), nvs_read_wifi_password());
```

Comment:
This is fine for now, but eventually we'll want to abstract this out into a common module that can be used by all samples. We'll also want to be able to select between using hardcoded or runtime settings, and between PSKs and certificates. Similar to `sample_credentials.h` that we use for the Zephyr examples (implemented by `examples/zephyr/common/hardcoded_credentials.c` and `examples/zephyr/common/settings_golioth.c`). I added https://github.com/golioth/firmware-issue-tracker/issues/429 for this work.